capstone/src/main/java/com/google/sps/data/RecommenderAlgorithm.java

```diff
@@ -11,49 +13,158 @@
+                !rooms.contains(room.getRoomKey())) {
+                rooms.add(room.getRoomKey());
+            }
+        }
+        return rooms;
     }
- 
-    public static HashMap<String, Integer> populateUserTagMap(User user) {
+
+    public static void populateEmbeddedEntity(User user) {
         DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-        Filter userRoomFilter = new FilterPredicate(Room.FOLLOWERS_PROPERTY_KEY, FilterOperator.EQUAL, user.getUserKey());
-        Query userRoomQuery = new Query(Room.ROOM_ENTITY_NAME).setFilter(userRoomFilter);
-        PreparedQuery userRoomResults = datastore.prepare(userRoomQuery);
- 
-        HashMap<String, Integer> userTagMap = Tag.constructUserVectorMap();
-        List<String> userTags = null;
-        for(Entity roomEntity : userRoomResults.asIterable()){
-            userTags = new Room(roomEntity).getAllTags();
-            for(String tag : userTags) {
-                userTagMap.put(tag, userTagMap.get(tag) + 1);
+        Query query = new Query(Room.ROOM_ENTITY_NAME);
+        PreparedQuery results = datastore.prepare(query);
+        EmbeddedEntity entity = user.getEmbeddedTags();
+
+        if (results.countEntities() == 0) {
+            return;
+        }
+
+        entity = Tag.constructUserEmbeddedEntity();
+        for (Entity result : results.asIterable()) {
+            Room room = new Room(result);
+            if (room.getAllFollowers().contains(user.getUserKey())) {
+                for (String tag : room.getAllTags()) {
+                    double frequency = (Double) entity.getProperty(tag);
+                    entity.setProperty(tag, frequency + 1.0);
+                }
             }
         }
+        user.setEmbeddedTags(entity);
+    }
+
+    public static void normalizeEmbeddedEntity(User user) {
+        RecommenderAlgorithm.populateEmbeddedEntity(user);
+        EmbeddedEntity entity = user.getEmbeddedTags();
+        double magnitude = RecommenderAlgorithm.magnitude(entity);
+        for (String tag : Tag.CATEGORY_TAGS) {
+            double value = (Double) entity.getProperty(tag);
+            if (magnitude != 0.0) {
+                entity.setProperty(tag, value/magnitude);
+            }
+        }
+        user.setEmbeddedTags(entity);
+    }
+
+    public static void normalizeAllEmbeddedEntities() {
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Query query = new Query(User.USER_ENTITY_NAME);
+        PreparedQuery results = datastore.prepare(query);
 
-        return userTagMap;
+        for (Entity entity : results.asIterable()) {
+            try{
+                RecommenderAlgorithm.normalizeEmbeddedEntity(DatabaseService.getUser(entity.getKey()));
+            } catch (EntityNotFoundException e) {
+                System.out.println("Entity not found!");
+            }
+        }
     }
+
```

Comment:
As a general rule methods should go in the order of dependencies. So since `normalizeEmbeddedEntity` is only used by this (I think), it should go after this method. If you think of it like reading a book, I am going top to bottom, and when I get to this method Im thinking "what? what is this for?" if I dont have the context for what kind of what it will be used.